extension-androidauto/src/main/java/com/mapbox/maps/extension/androidauto/BaseWidgetRenderer.kt

```diff
@@ -0,0 +1,393 @@
+  private val marginRight: Float = 0f,
+  private val marginBottom: Float = 0f,
+) : BaseWidget() {
+
+  private var program = 0
+  private var vertexPositionHandle = 0
+  private var texturePositionHandle = 0
+  private var textureHandle = 0
+  private var screenMatrixHandle = 0
+  private var rotationMatrixHandle = 0
+  private var translateMatrixHandle = 0
+  private val textures = intArrayOf(0)
+
+  private var vertexShader = 0
+  private var fragmentShader = 0
+
+  private var heightOffset: Float = 0f
+  private var widthOffset: Float = 0f
+
+  private lateinit var screenMatrixData: FloatArray
+  private lateinit var screenMatrixBuffer: FloatBuffer
+
+  private val rotationMatrixData = FloatArray(16).apply {
+    Matrix.setIdentityM(this, 0)
+  }
+  private var rotationDirty: Boolean = true
+  private lateinit var rotationMatrixBuffer : FloatBuffer
+
+  private val translateMatrix = FloatArray(16).apply {
+    Matrix.setIdentityM(this, 0)
+  }
+  private var translationDirty: Boolean = true
+  private lateinit var translationMatrixBuffer : FloatBuffer
+
+  private fun getMatrixBuffer(matrixData: FloatArray): FloatBuffer {
+    return ByteBuffer.allocateDirect(matrixData.size * BYTES_PER_FLOAT).run {
+      // use the device hardware's native byte order
+      order(ByteOrder.nativeOrder())
+
+      // create a floating point buffer from the ByteBuffer
+      asFloatBuffer().apply {
+        // add the coordinates to the FloatBuffer
+        put(matrixData)
+        // set the buffer to read the first coordinate
+        rewind()
+      }
+    }
+  }
+
+  private lateinit var vertexPositionData: FloatArray
+  private lateinit var vertexPositionBuffer: FloatBuffer
+
+  private val texturePositionData = floatArrayOf(
+    0f, 0f,
+    0f, 1f,
+    1f, 0f,
+    1f, 1f
+  )
+  private val texturePositionBuffer: FloatBuffer by lazy {
+    getMatrixBuffer(texturePositionData)
+  }
+
+  override var dirty: Boolean = true
+
+  override fun onSizeChanged(width: Int, height: Int) {
+    super.onSizeChanged(width, height)
+    heightOffset = when (position) {
+      WidgetPosition.BOTTOM_LEFT -> height.toFloat() - bitmap.height.toFloat() / 2f - marginBottom
+      WidgetPosition.BOTTOM_RIGHT -> height.toFloat() - bitmap.height.toFloat() / 2f - marginBottom
+      WidgetPosition.TOP_LEFT -> marginTop + bitmap.height.toFloat() / 2f
+      WidgetPosition.TOP_RIGHT -> marginTop + bitmap.height.toFloat() / 2f
+    }
+    widthOffset = when (position) {
+      WidgetPosition.TOP_RIGHT -> width.toFloat() - bitmap.width.toFloat() / 2f - marginRight
+      WidgetPosition.BOTTOM_RIGHT -> width.toFloat() - bitmap.width.toFloat() / 2f - marginRight
+      WidgetPosition.TOP_LEFT -> marginLeft + bitmap.width.toFloat() / 2f
+      WidgetPosition.BOTTOM_LEFT -> marginLeft + bitmap.width.toFloat() / 2f
+    }
+
+    Matrix.translateM(
+      translateMatrix,
+      0,
+      widthOffset,
+      heightOffset,
+      0f
+    )
+
+    // The uScreen matrix
```

Comment:
This comment is huge, perhaps we could shorten it.

Reply:
one-liner now